### PR TITLE
Add provenance generation tool

### PR DESCRIPTION
**Pull request by joinemm**
Adds python script to generate provenance file according to SLSA v1 specification, and documentation for it.

### CODE REVIEW

File: doc/provenance.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+To generate provenance file for `curl-8.6.0` in your nix store:
+
+```bash
+provenance /nix/store/fh7vxc5xgiwl6z7vwq5c3lj84mpcs4br-curl-8.6.0-bin
+```
+
+By default the dependecies are resolved only at the top level. ie. only direct dependecies.
+To get all dependencies recursively, you can use the `--recursive` option.
+Note the this will result in a very long provenance file.
```
Comment (henrirosten) [inline doc/provenance.md:37]: I would add a mention somewhere here that the dependencies now included are buildtime depdendencies.
I don't know if SLSA specs actually state if the provenance should include buildtime or only runtime dependencies.

Not necessary a topic of this PR, but if you are going to add a support to outputting only the runtime dependencies, you should align the command-line usage with other tools in this repo, that is, default to runtime dependencies with the `--buildtime` option changing the output to include also buildtime dependencies.

File: README.md [hunk 2]
```diff
@@ -102,6 +102,9 @@ $ src/repology/repology_cve.py --help
 
 # nix_outdated:
 $ src/nixupdate/nix_outdated.py --help
+
+# provenance:
+$ src/provenance/main.py --help
```
Comment (henrirosten) [inline README.md:107]: You could also add a link to the provenance.md on this README among the other listed tools section: https://github.com/tiiuae/sbomnix/blob/main/README.md?plain=1#L10-L15

**Review by henrirosten — COMMENTED** [review]
In addition to the comments below, please consider adding at least some basic test cases. 
Perhaps at least a test case that checks the provenance-tool generates output json that matches the expected schema? Perhaps something similar to: https://github.com/tiiuae/sbomnix/blob/9df4fb380b94fce54c4374958e7fc69a573cf6ec/tests/test_sbomnix.py#L107-L128

File: doc/provenance.md (hunk 1 above)
Comment (joinemm) [reply doc/provenance.md:37]: from [SLSA spec](https://slsa.dev/spec/v1.0/provenance): 
> All artifacts fetched during initialization or execution of the build process are considered dependencies

To me this sounds like only buildtime dependencies are relevant.

File: doc/provenance.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+PROVENANCE_TIMESTAMP_FINISHED="$(date +%s)"
+
+PROVENANCE_EXTERNAL_PARAMETERS="$(jq -n --arg target "$target" '$ARGS.named')"
+PROVENANCE_INTERNAL_PARAMETERS="$(jq -n --arg nixVersion $(nix --version) '$ARGS.named')"
+
+export PROVENANCE_TIMESTAMP_BEGIN
+export PROVENANCE_TIMESTAMP_END
+export PROVENANCE_EXTERNAL_PARAMETERS
+export PROVENANCE_INTERNAL_PARAMETERS
```
[the remark this answers is not in the document]
Comment (joinemm) [reply doc/provenance.md:77]: good catch

**Review by henrirosten — APPROVED** [review]

**Pull request merged**